### PR TITLE
feat(skore): Add an HTML representation for interactive environment

### LINKED ISSUE #1104: feat(skore): create an HTML representation for `.help` in the differen reports
Right now, `EstimatorReport.help` shows a tree representation of all available functionalities available for the reporter using `rich`. However, it will become overwhelming at the time that we add more and more accessors.

Therefore, I think that we should still have `rich` backend whenever we are in a terminal but we should come with a richer representation with dropdown menu whenever we are in a notebook or IDE environment where we can leverage HTML. It will the UI/UX more enjoyable. I can also foresee to be able to click on a specific method to copy and then paste the access to a specific method of an accessor.

I'm thinking that we could then reduce the amount of information that we display with rich: never go beyond 1 level but make the help available at each level. The dynamic aspect of the HTML makes that it will be easier to handle.

### PR DESCRIPTION
**Pull request by glemaitre**
closes #1104 

This is an experimental PR to show what could be done regarding having an HTML representation when we are in an interactive environment.

Check the documentation preview and the getting started example to see the rendering and behaviour.

This PR is doing a full refactoring of the `repr` module:

- we are creating a module specifically to handle rich representation
- we are creating a module specifically to handle HTML representation
- we have have jinja template for the html representation
- we create a data module to define what are the models used to inject the information into the rich and HTML representation. It contains as well all the utility to create those data model
- we create a base module containing a report, display, and accessor representation that delegate to the rich or HTML representation.
- finally, we tested each individual module which make it easier to ensure the proper behavior instead of an integrated tests for each report.

### CODE REVIEW

File: skore/src/skore/_sklearn/_base.py [hunk 1]
```diff
@@ -18,8 +19,648 @@
...
+
+        return Panel(
+            content,
+            title=self._get_help_panel_title(),
+            expand=False,
+            border_style="orange1",
+        )
+
+
+class _HTMLHelpMixin(ABC):
```
Comment (rouk1) [inline skore/src/skore/_sklearn/_base.py]: 😿 

Did you consider using an alternative approach that may be more maintainable ?
Two track to consider:
- a web component that gets hydrated by a json model 
- a templating language like Jinja (it is already available as a transient dependency through skrub iirc)

Both those strategy would avoid fragile escaping that made my eyes bleed 😆

Comment (glemaitre) [reply skore/src/skore/_sklearn/_base.py]: Yes I think it is reasonable. It was to check the visualization but the solution is horrible to maintain.

I can think easily of a jinja. Do you have a resource where I could look at the first solution.

**Comment by glemaitre** [comment]
I think it is good to go actually.

File: sphinx/Makefile [hunk 2]
```diff
@@ -6,6 +6,7 @@ ifeq ($(SPHINX_DOMAIN),)
 	SPHINX_DOMAIN := docs.skore.probabl.ai
 endif
 
+export SPHINX_BUILD ?= 1
```
Comment (thomass-dev) [inline sphinx/Makefile:9]: Take a look on the line 17, isn't in conflict?

File: skore/src/skore/_utils/repr/help.css [hunk 3]
```diff
@@ -0,0 +1,292 @@
+/* Light theme (default) */
+.skore-help-container,
```
Comment (rouk1) [inline skore/src/skore/_utils/repr/help.css]: As you're using a shadow dom, you do not have to prefix all your class names. They will not leak to the host page.
Maybe you could only use semantic class name for readability.

File: skore/src/skore/_utils/repr/display_help-content.html.j2 [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<div class="skore-help-container">
+    <div class="skore-help-title">{{ title }}</div>
```
Comment (rouk1) [inline skore/src/skore/_utils/repr/display_help-content.html.j2]:
```suggestion
    <header class="skore-help-title">{{ title }}</header>
```

File: skore/src/skore/_utils/repr/display_help-content.html.j2 [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            <span class="skore-help-label">{{ root_node }}</span>
+        </div>
+        <div class="skore-help-root-content">
+            {# Methods section #}
+            {% if methods %}
+                <div class="skore-help-node skore-help-folder">
+                    <input type="checkbox" id="{{ methods_section.id }}" class="skore-help-toggle">
+                    <label for="{{ methods_section.id }}" class="skore-help-folder-label">
+                        <span class="skore-help-section-name">Methods</span>
+                    </label>
```
Comment (rouk1) [inline skore/src/skore/_utils/repr/display_help-content.html.j2]: You can avoid an id by nesting the input inside the label.


```suggestion
                    <label class="skore-help-folder-label">
                        <input type="checkbox" class="skore-help-toggle">
                        <span class="skore-help-section-name">Methods</span>
                    </label>
```

File: skore/src/skore/_utils/repr/display_help-content.html.j2 [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                    <div class="skore-help-folder-content" id="{{ methods_section.branch_id }}">
+                        {% for method in methods %}
+                            <div class="skore-help-node skore-help-method">
+                                <span class="skore-help-method-name">.<a href="{{ method.doc_url }}" target="_blank" class="skore-help-method-tooltip">{{ method.name }}<span class="skore-help-tooltip-text">{{ method.description|e }}<br><br>Click to access the online documentation</span></a>{{ method.parameters }}</span>
```
Comment (rouk1) [inline skore/src/skore/_utils/repr/display_help-content.html.j2]: `<br/>` should in my opinion be avoid as much as possible. Formatting should happen at the CSS level.
Here you create a structure that could have been clearly represented with tags.

Also it is not 1998 anymore "click here" is a double information, your link is already in a anchor tag, user will understand that 😅

File: skore/src/skore/_utils/repr/display_help-content.html.j2 [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            {% if attributes %}
+                <div class="skore-help-node skore-help-folder">
+                    <input type="checkbox" id="{{ attributes_section.id }}" class="skore-help-toggle">
+                    <label for="{{ attributes_section.id }}" class="skore-help-folder-label">
+                        <span class="skore-help-section-name">Attributes</span>
+                    </label>
+                    <div class="skore-help-folder-content" id="{{ attributes_section.branch_id }}">
+                        {% for attr in attributes %}
+                            <div class="skore-help-node skore-help-method">
+                                <span class="skore-help-method-name">.<a href="{{ attr.doc_url }}" target="_blank" class="skore-help-method-tooltip">{{ attr.name }}<span class="skore-help-tooltip-text">{{ attr.description|e }}<br><br>Click to access the online documentation</span></a></span>
```
Comment (rouk1) [inline skore/src/skore/_utils/repr/display_help-content.html.j2]: Same as above.

File: sphinx/Makefile (hunk 2 above)
Comment (glemaitre) [reply sphinx/Makefile:9]: Nop, it is an environment variable to mention that we are building the documentation.

The one below is used by the script as an internal variable.

File: skore/src/skore/_utils/repr/display_help-content.html.j2 (hunk 6 above)
Comment (glemaitre) [reply skore/src/skore/_utils/repr/display_help-content.html.j2]: Let's remove the part of the code where I'm clearly demonstrated that I'm old fashion ;)

**Comment by glemaitre** [comment]
I addressed the comment. I'll modify the documentation of accessor to be closer to the same pattern than report to be able to fold the methods. It will be more consistent.

**Comment by glemaitre** [comment]
It is good on my side. You can have a look at the generated documentation to have a preview.
In another PR, I could reuse the HTML generated code to add it in the landing page.

**Comment by glemaitre** [comment]
> I played with it a little in the docs preview and noticed that for CrossValidationReports and ComparisonReports, for non linear models, the feature_importance accessor appears but can't be interacted with since there is no available feature importance method in skore in these cases yet. Is it on purpose ? Shouldn't we rather hide the accessor if it can access nothing ?

The accessor is actually register programmatically but do not have method. I did not think about this case. I'll make sure that we append the accessor only if at least an item is available. If it makes it difficult then, we can just wait because it will bee soonish available.

**Comment by glemaitre** [comment]
@GaetandeCast It was quite trivial to manage this corner case by avoiding adding an accessor if we did not have any method.

I added some non-regression test.

File: skore/src/skore/_utils/repr/display_help.html.j2
Comment (auguste-probabl) [inline skore/src/skore/_utils/repr/display_help.html.j2:1]: Since we depend explicitly on jinja2 now it should be added to pyproject.toml and the lock files should be regenerated

Comment (glemaitre) [reply skore/src/skore/_utils/repr/display_help.html.j2:1]: I added the dependencies and generated the lock files.

File: skore/src/skore/_utils/repr/display_help.html.j2 [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    const template = document.getElementById('{{ container_id }}-template');
+    if (!template) {
+        return;
+    }
+
+    // Create shadow root with closed mode for better isolation
+    const shadowRoot = container.attachShadow({ mode: 'closed' });
+    shadowRoot.appendChild(template.content.cloneNode(true));
```
Comment (auguste-probabl) [inline skore/src/skore/_utils/repr/display_help.html.j2:31]: For the record, I still think this is an anti-pattern and we can achieve isolation with some simple prefixing of attributes (e.g. call every class `skore-help-*`). For instance, this could help us get rid of the Javascript code that just checks for dark mode: dark mode is now natively supported in CSS via `@media (prefers-color-scheme: dark)`

File: skore/src/skore/_utils/repr/help.css [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+    --color-blue: #6b42f6;
+    --color-orange: #f97316;
+    --color-text: #fff;
+    --color-background: #000;
+    --color-orange-hover: rgba(249, 115, 22, 0.2);
+    --color-blue-hover: rgba(107, 66, 246, 0.2);
+}
+
+.container {
+    font-family: 'Monaco', 'Menlo', 'Ubuntu Mono', 'Consolas', monospace;
```
Comment (auguste-probabl) [inline skore/src/skore/_utils/repr/help.css]: I'd prefer `monospace` to be the default, because then the user has control over it.

File: skore/src/skore/_utils/repr/html_repr.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+from jinja2 import Environment, FileSystemLoader
+
+from skore._utils.repr.data import (
+    _AccessorHelpDataMixin,
+    _DisplayHelpDataMixin,
+    _ReportHelpDataMixin,
+)
+
+
+def get_jinja_env():
```
Comment (auguste-probabl) [inline skore/src/skore/_utils/repr/html_repr.py:17]: This could be cached

File: skore/src/skore/_utils/repr/rich_repr.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+        base_methods = data.get("base_methods", [])
+        if base_methods:
+            methods_branch = tree.add("[bold cyan]Methods[/bold cyan]")
+            for method in base_methods:
+                displayed_name = f"{method['name']}(...)"
+                description = method["description"]
+                methods_branch.add(f".{displayed_name}".ljust(26) + f" - {description}")
+
+        attributes = data.get("attributes") or []
```
Comment (auguste-probabl) [inline skore/src/skore/_utils/repr/rich_repr.py]:
```suggestion
        attributes = data.get("attributes", [])
```

File: skore/tests/unit/reports/estimator/feature_importance/test_accessor.py [hunk 12]
```diff
@@ -1,3 +1,5 @@
+import pytest
```
Comment (auguste-probabl) [inline skore/tests/unit/reports/estimator/feature_importance/test_accessor.py]: Unused import?

File: skore/tests/unit/utils/repr/test_base.py [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    display_with_base_help.help()
+    mock_notebook.assert_called_once()
+    mock_html_cls.assert_called_once()
+    html_arg = mock_html_cls.call_args[0][0]
+    assert "skore-display-help-" in html_arg
+    mock_display.assert_called_once_with(sentinel_html)
+
+
+@patch("skore._utils.repr.base.is_environment_notebook_like", return_value=False)
+@patch("skore.console.print")
```
Comment (auguste-probabl) [inline skore/tests/unit/utils/repr/test_base.py]: Nice; in the rest of the test suite we tend to use the pytest `monkeypatch` fixture but since this is builtin, I would tend to prefer it. We should eventually pick one and be consistent.

File: skore/tests/unit/utils/repr/test_data.py [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+    ],
+)
+def test_build_attribute_text_fragment(obj, attribute_name, expected):
+    """_build_attribute_text_fragment returns encoded name,-type or encoded name."""
+    got = _build_attribute_text_fragment(obj, attribute_name)
+    if expected is not None:
+        expected = f"{quote(attribute_name, safe='')},-{quote(expected, safe='')}"
+    else:
+        expected = quote(attribute_name, safe="")
+    assert got == expected
```
Comment (auguste-probabl) [inline skore/tests/unit/utils/repr/test_data.py]:
```suggestion
@pytest.mark.parametrize(
    "obj, attribute_name, expected",
    [
        (_ClassWithNumpydocAttrs(), "coef_", "coef_,-ndarray (n_features,)"),
        (_ClassWithNumpydocAttrs(), "param", "param_,-int"),
        (_ClassWithNoDocstring(), "coef_", "coef_"),
        (_ClassWithNumpydocAttrs(), "other_attr", "other_attr"),
    ],
)
def test_build_attribute_text_fragment(obj, attribute_name, expected):
    """_build_attribute_text_fragment returns encoded name,-type or encoded name."""
    assert _build_attribute_text_fragment(obj, attribute_name) == expected
```

File: skore/tests/unit/utils/repr/test_base.py (hunk 13 above)
Comment (thomass-dev) [reply skore/tests/unit/utils/repr/test_base.py]: Just a general note: prefer `pytest` when possible, it is used all over the repository.
We can reopen the debate `unittest` or `pytest`, but i don't think it is the right place 😄 .

File: skore/tests/unit/utils/repr/test_data.py [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+    "method_name, expected",
+    [
+        ("documented", "First line of docstring."),
+        ("undocumented", "No description available"),
+    ],
+)
+def test_get_method_short_summary(method_name, expected):
+    """get_method_short_summary returns first docstring line or fallback."""
+    obj = _ClassWithDocstringMethod()
+    method = getattr(obj, method_name)
```
Comment (auguste-probabl) [inline skore/tests/unit/utils/repr/test_data.py]:
```suggestion
    method = getattr(_ClassWithDocstringMethod(), method_name)
```

File: skore/tests/unit/utils/repr/test_base.py (hunk 13 above)
Comment (thomass-dev) [reply skore/tests/unit/utils/repr/test_base.py]: For the specific case when you need to assert that your mock has been called, combining `pytest.monkeypatch.setattr`  and `unittest.mock.Mock` is a good solution.

File: skore/tests/unit/utils/repr/test_data.py [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+
+
+def test_get_documentation_url_class_only(report_with_methods, display_with_methods):
+    """get_documentation_url returns base class URL with no accessor/method/attribute."""
+    for obj in (report_with_methods, display_with_methods):
+        url = get_documentation_url(obj=obj)
+        assert url.startswith("https://docs.skore.probabl.ai/")
+        assert "/reference/api/" in url
+        assert "skore." in url and obj.__class__.__name__ in url
+        assert url.endswith(".html")
```
Comment (auguste-probabl) [inline skore/tests/unit/utils/repr/test_data.py]: Nit
```suggestion
        assert re.match(rf'https://docs.skore.probabl.ai/[^/]+/reference/api/skore\.[^/]+{obj.__class__.__name__}.html', url)
```

File: skore/tests/unit/utils/repr/test_html_repr.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+@pytest.fixture
+def accessor_with_html(report_with_html):
+    """Accessor with HTML help mixin."""
+    return _AccessorWithHTML(parent=report_with_html)
+
+
+@pytest.fixture
+def display_with_html():
+    """Display with HTML help mixin."""
+    return _DisplayWithHTML()
```
Comment (auguste-probabl) [inline skore/tests/unit/utils/repr/test_html_repr.py]: Inline those

File: sphinx/Makefile (hunk 2 above)
Comment (auguste-probabl) [reply sphinx/Makefile:9]: They are pretty similar which is confusing. Rename maybe to `IS_SPHINX_BUILD`, or even better: don't define a new variable and instead check for `SPHINX_DOMAIN`, `SPHINX_VERSION` or `SPHINX_RELEASE`

File: skore/src/skore/_utils/repr/data.py [hunk 18]
```diff
@@ -0,0 +1,543 @@
...
+        sig = inspect.signature(method)
+        parameter_names = [
+            parameter_name
+            for parameter_name, _ in sig.parameters.items()
+            if parameter_name != "self"
+        ]
+        if parameter_names:
+            parameters = "(" + ", ".join(parameter_names) + ")"
+        else:
+            parameters = "()"
```
Comment (auguste-probabl) [inline skore/src/skore/_utils/repr/data.py]:
```suggestion
        parameters = f"({", ".join(parameter_names)})"
```

File: skore/src/skore/_utils/repr/data.py [hunk 19]
```diff
@@ -0,0 +1,543 @@
...
+
+        description = get_method_short_summary(method)
+        favorability = None
+        if hasattr(obj, "_get_favorability_text"):
+            favorability = obj._get_favorability_text(name)
```
Comment (auguste-probabl) [inline skore/src/skore/_utils/repr/data.py]:
```suggestion
        favorability = obj._get_favorability_text(name) if hasattr(obj, "_get_favorability_text") else None
```

File: skore/src/skore/_utils/repr/display_help.html.j2 (hunk 8 above)
Comment (glemaitre) [reply skore/src/skore/_utils/repr/display_help.html.j2:31]: We exactly did way to avoid prepanding using the shadow-DOM. Here, I would still stick to this anti-pattern because the usual pattern troubled us so much before.

File: skore/src/skore/_utils/repr/data.py [hunk 20]
```diff
@@ -0,0 +1,543 @@
...
+            AttributeHelp(
+                name=attr_name,
+                description=get_attribute_short_summary(self, attr_name),
+                doc_url=get_documentation_url(
+                    obj=self,
+                    attribute_name=attr_name,
+                ),
+            )
+            for attr_name in attrs_without_underscore + attrs_with_underscore
+        ]
```
Comment (auguste-probabl) [inline skore/src/skore/_utils/repr/data.py]:
```suggestion
        items = [
            AttributeHelp(
                name=attr_name,
                description=get_attribute_short_summary(self, attr_name),
                doc_url=get_documentation_url(obj=self, attribute_name=attr_name),
            )
            for attr_name in attributes
        ]
```

**Review by auguste-probabl — CHANGES_REQUESTED** [review]
Nearly there. There are some more deep stuff like the shadow DOM thing, that it would be nice to go over again. Otherwise it's looking good.

File: skore/src/skore/_utils/repr/display_help.html.j2 (hunk 8 above)
Comment (auguste-probabl) [reply skore/src/skore/_utils/repr/display_help.html.j2:31]: What's wrong with prepending? If it's just an issue of adding noise, I think it's worth it given the complexity that the shadow DOM brings.

File: skore/src/skore/_utils/repr/html_repr.py (hunk 10 above)
Comment (glemaitre) [reply skore/src/skore/_utils/repr/html_repr.py:17]: If templating with `jinja` comes somewhere else then having a function local to the module is better because we load things relatively to it.

File: skore/tests/unit/reports/estimator/feature_importance/test_accessor.py (hunk 12 above)
Comment (glemaitre) [reply skore/tests/unit/reports/estimator/feature_importance/test_accessor.py]: Indeed. I messed up the merge conflict.

File: skore/tests/unit/utils/repr/test_base.py (hunk 13 above)
Comment (glemaitre) [reply skore/tests/unit/utils/repr/test_base.py]: Let's keep `pytest` :)

File: skore/src/skore/_utils/repr/display_help.html.j2 (hunk 8 above)
Comment (rouk1) [reply skore/src/skore/_utils/repr/display_help.html.j2:31]: > dark mode is now natively supported in CSS via @media (prefers-color-scheme: dark)

This is true but there is also IDE (like vs code and its forks and Jupyter) which use different strategies to select themes. 

Also event if I agree that shadow dom adds complexity it also comes with greta isolation power which is nice especially in a notebook.

**Comment by glemaitre** [comment]
Could we merge this PR :)
